### DISCUSSION TITLE
Helix Integration: Plugin help to write hovered file path to env variable and then exit

### DISCUSSION THREAD

**Discussion by RoloEdits** [sxyazi/yazi#2096] [1. Q&A]
### What system are you running Yazi on?

Windows

### What terminal are you running Yazi in?

wezterm

### `yazi --debug` output

```Shell
Yazi
    Version: 0.4.2 (5cfcab3 2024-12-20)
    Debug  : false
    Triple : x86_64-pc-windows-msvc (windows-x86_64)
    Rustc  : 1.83.0 (90b35a62 2024-11-26)

Ya
    Version: 0.4.2 (5cfcab3 2024-12-20)

Emulator
    Brand.from_env      : Some(WezTerm)
    Emulator.detect     : Emulator { kind: Left(WezTerm), light: false, cell_size: None }
    Emulator.detect_full: Ok(Emulator { kind: Right(Unknown { kgp: false, sixel: false }), light: false, cell_size: None })

Adapter
    Adapter.matches: Iip

Desktop
    XDG_SESSION_TYPE           : None
    WAYLAND_DISPLAY            : None
    DISPLAY                    : None
    SWAYSOCK                   : None
    HYPRLAND_INSTANCE_SIGNATURE: None
    WAYFIRE_SOCKET             : None

SSH
    shared.in_ssh_connection: false

WSL
    WSL: false

NVIM
    NVIM          : false
    Neovim version: program not found

Variables
    SHELL           : None
    EDITOR          : Some("hx")
    VISUAL          : None
    YAZI_FILE_ONE   : Some("C:\\Program Files\\Git\\usr\\bin\\file.exe")
    YAZI_CONFIG_HOME: None
    YAZI_ZOXIDE_OPTS: None
    FZF_DEFAULT_OPTS: None

Text Opener
    default     : Some(Opener { run: "code %*", block: false, orphan: true, desc: "code", for_: None, spread: true })
    block-create: Some(Opener { run: "code -w %*", block: true, orphan: false, desc: "code (block)", for_: None, spread: true })
    block-rename: Some(Opener { run: "code -w %*", block: true, orphan: false, desc: "code (block)", for_: None, spread: true })

Multiplexers
    TMUX               : 0
    tmux version       : program not found
    tmux build flags   : enable-sixel=Unknown
    ZELLIJ_SESSION_NAME: None
    Zellij version     : program not found

Dependencies
    file          : 5.45
    ueberzugpp    : program not found
    ffmpeg/ffprobe: 7.1 / 7.1
    pdftoppm      : program not found
    magick        : program not found
    fzf           : 0.56.3
    fd/fdfind     : 10.2.0 / program not found
    rg            : 14.1.1
    chafa         : program not found
    zoxide        : 0.9.6
    7z/7zz        : 24.09 / program not found
    jq            : program not found

Clipboard
    wl-copy/paste: program not found / program not found
    xclip        : program not found
    xsel         : program not found
```

### Describe the question

I am developing https://github.com/helix-editor/helix/pull/12288 for helix that hopes to allow something like:
```toml
[keys.normal.space]
e = [":sh --no-popup wezterm cli spawn --floating-pane yazi %{path}", ":open --env YAZI_LAST_OPENED_FILE"]
```

This would allow for a more integrated experience using yazi as a file tree. 

The main feature that I am trying to make with the plugin is to get the currently hovered file, and when the `open` action done (in this case just a keypress), instead of opening like normal, it will set the environment variable and exit. Its needed to exit as `:open` wont run until the wezterm pane closes. Then `:open` reads from the environment variable.

(`%{path}` is a placeholder for  variable expansion, which is also in the works, which expands to the current buffers path, which when opening yazi will do so at)

I am testing out features and would love to be able to demo this. I have looked at some plugins as examples but cant even seem to use them when installed with `ya`.

Thanks!

### Checklist

- [X] I have searched the existing discussions/issues

**Comment by AnirudhG07** [comment]
Can you share the recent ` ~/.local/state/yazi/yazi.log` outputs?
Also which plugins are you trying? Can you share a bit more info on it? Are you trying to create a plugin as well with lua for testing?

**Reply by RoloEdits** [reply]
> Can you share the recent  ~/.local/state/yazi/yazi.log outputs?

Where would this be on windows? 

> Also which plugins are you trying? Can you share a bit more info on it?

yes, I was trying to use https://github.com/yazi-rs/plugins/tree/main/smart-enter.yazi
It seemed pretty close to what I wanted to do, so I was looking at it as a jumping off point. 

```yoml
[[manager.prepend_keymap]]
on   = "<Enter>"
run  = "plugin smart-enter"
desc = "Enter the child directory, or open the file"
```

This is the structure of the `yazi/` config directory:

```shell
├───config
│   └───plugins
│       └───smart-enter.yazi
└───state
    └───packages
        └───d3809db7f978e2fb3d48192ce292a622
            ├───.github
            │   ├───DISCUSSION_TEMPLATE
            │   ├───ISSUE_TEMPLATE
            │   └───workflows
            ├───chmod.yazi
            ├───diff.yazi
            ├───full-border.yazi
            ├───git.yazi
            ├───hide-preview.yazi
            ├───jump-to-char.yazi
            ├───lsar.yazi
            ├───mactag.yazi
            ├───max-preview.yazi
            ├───mime-ext.yazi
            ├───no-status.yazi
            ├───smart-enter.yazi
            ├───smart-filter.yazi
            └───sudo-demo.yaz
```

It was installed with `scoop`.


> Are you trying to create a plugin as well with lua for testing?

Yes, exactly. I looked to `smart-paste` for inspiration as well, but again can't get it to work. 

The idea is to merge the `smart-enter` concept of merging `enter` and `open`, where here pressing `ENTER` would enter a directory, but if it was  a file it would write the path the an environment variable and then `close`.

This is what I was trying to get working, but also to no avail.
```lua
return {
  entry = function(self)
    local hov = cx.active.current.hovered
    if hov then
      if hov.is_dir and "enter" or "open" then
        ya.manager_emit("enter", {})
      elseif "open" then
        os.setenv("YAZI_LAST_FILE_OPENED", hov.path)
        ya.manager_emit("close", {})
      end
    end
  end
}
```

Thanks once again.

**Reply by AnirudhG07** [reply]
In your code for line 'hov.is_dir.....' I am not sure what is equal to "enter" or "open". Are you missing job.args maybe?
Also I hope you are not running 'lua plugin.lua' in terminal since it's meant to be run as plugin through keybindings(ig uk this).

**Reply by RoloEdits** [reply]
> In your code for line 'hov.is_dir.....' I am not sure what is equal to "enter" or "open". Are you missing job.args maybe?

I'm not sure either, this was me trying to fumble around using `smart-enter` as a base:

```lua
-- `smart-enter.yazi/init.lua`
local function setup(self, opts) self.open_multi = opts.open_multi end

local function entry(self)
	local h = cx.active.current.hovered
	ya.manager_emit(h and h.cha.is_dir and "enter" or "open", { hovered = not self.open_multi })
end

return { entry = entry, setup = setup }
```

> Also I hope you are not running 'lua plugin.lua' in terminal since it's meant to be run as plugin through keybindings(ig uk this).

I was setting it as a keybind, like how `smart-enter` was, but with the new plugin name. I also took inspiration from the `smart-paste` example in the docs. 

I see some parts of the docs show there is supposed to be an `init.lua` at a higher level than the configs, but the unix parts show its at different place than the windows one. Was there supposed to be one auto-generated or was I supposed to also make one? Is there boilerplate I must add?

The instructions only mention using `ya` and then offer what seem like optional steps to have logic about when a plugin should be active or not (Like reading the vim variable to only be active in a vim context).

Sorry if these seem like basic questions, not familiar with lua or the plugin api.

**Reply by AnirudhG07** [reply]
I wont be able to ans for windows config str, 
But yeah ull have to make init.lua, and there is no particular boiler plate for it.
It's not a must requirement but many plugins require it. Your special configs for Yazi go in there.

**Comment by sxyazi** [comment]
> os.setenv("YAZI_LAST_FILE_OPENED", hov.path)

There is no `os.setenv()` in Lua, only [`os.getenv()`](https://www.lua.org/manual/5.4/manual.html#pdf-os.getenv), so this won't work.

Yazi has a `--chooser-file` option, which accepts a temporary file. When a user opens files in Yazi, it writes the paths of the opened files (either one or multiple) into that temporary file, separated by LF (`\n`). Does this work for you?

**Reply by RoloEdits** [reply]
Hmm, Ill have to play around more to see if I can get something to work. Is there a way to set the output as stdout on windows using `chooser-file`?

**Reply by sxyazi** [reply]
I don't think so, it seems that Windows doesn't have something like `/dev/stdout` so you can pass `chooser-file=/dev/stdout`.

If you really need stdout, maybe we can add a virtual path like `chooser-file=yazi:///dev/stdout` for Windows.

But I'd recommend using a temporary file, as it can be used together with other options (like `--cwd-file`, `--local-events` or `--remote-events`) to avoid mixing their outputs with stdout, making it hard to distinguish them.
